### ISSUE TITLE
cgroups with slurm are not read correctly

### ISSUE THREAD

**Issue by multimeric** [rstudio/rstudio#13690]
### System details

    RStudio Edition: Server
    RStudio Version :  2023.06.2 Build 561
    OS Version      : CentOS 7
    R Version       :  n/a

### Steps to reproduce the problem
* Run RStudio within Slurm, with a memory limit
* Open the Memory Usage Report
* Note that the free system memory will be listed as higher than the job's memory limit, and the "source" will be listed as `/proc/meminfo`

### Describe the problem in detail

RStudio does not seem to acknowledge Slurm's cgroups.

### Describe the behavior you expected

I expect RStudio to read the slurm cgroup information, and to therefore treat the memory limit as the same as `cat /sys/fs/cgroup/memory/slurm/uid_$UID/job_$SLURM_JOB_ID/memory.limit_in_bytes`

- [ ] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ ] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by multimeric** [comment]
One possibility is that `RStudio` is reading the cgroup location from `/proc/self/cgroup`, which isn't the relevant path, since that includes the task limits. e.g. `/slurm/uid_2687/job_13759490/step_batch/task_0`, whereas `/slurm/uid_2687/job_13759490/memory.limit_in_bytes` is the relevant cgroup for the entire job.

**Comment by ronblum** [comment]
@multimeric Thank you for raising this! A couple of questions:
1. Does it look like what's going on in issue #10928?
2. Is this in the open source RStudio Server, or are you using RStudio Server Pro/Posit Workbench?

**Comment by multimeric** [comment]
Yeah I had a look at #10928 before posting this, but I'm not sure it's exactly the same issue. It might be, because they both seem to fall back on `/proc/meminfo`, but unlike the kubernetes example, the concatenation of `/sys/fs/cgroup/memory` and the contents of `/proc/self/cgroup` do seem to exist, they're just the wrong numbers to use because the job cgroup, not the task cgroup is the relevant one.

This is with open source RStudio Server.

**Comment by multimeric** [comment]
Here are some relevant cgroup outputs:

```bash
$ cat /proc/self/cgroup 
11:blkio:/system.slice/slurmd.service
10:devices:/slurm/uid_3119/job_13759394/step_batch/task_0
9:cpuset:/slurm/uid_3119/job_13759394/step_batch
8:hugetlb:/
7:cpuacct,cpu:/system.slice/slurmd.service
6:freezer:/slurm/uid_3119/job_13759394/step_batch
5:memory:/slurm/uid_3119/job_13759394/step_batch/task_0
4:perf_event:/
3:pids:/system.slice/slurmd.service
2:net_prio,net_cls:/
1:name=systemd:/system.slice/slurmd.service
$ ls /sys/fs/cgroup/memory/slurm/uid_3119/job_13759394/step_batch/task_0
cgroup.clone_children           memory.kmem.tcp.max_usage_in_bytes  memory.oom_control
cgroup.event_control            memory.kmem.tcp.usage_in_bytes      memory.pressure_level
cgroup.procs                    memory.kmem.usage_in_bytes          memory.soft_limit_in_bytes
memory.failcnt                  memory.limit_in_bytes               memory.stat
memory.force_empty              memory.max_usage_in_bytes           memory.swappiness
memory.kmem.failcnt             memory.memsw.failcnt                memory.usage_in_bytes
memory.kmem.limit_in_bytes      memory.memsw.limit_in_bytes         memory.use_hierarchy
memory.kmem.max_usage_in_bytes  memory.memsw.max_usage_in_bytes     notify_on_release
memory.kmem.slabinfo            memory.memsw.usage_in_bytes         tasks
memory.kmem.tcp.failcnt         memory.move_charge_at_immigrate
memory.kmem.tcp.limit_in_bytes  memory.numa_stat
$ cat /sys/fs/cgroup/memory/slurm/uid_3119/job_13759394/step_batch/task_0/memory.limit_in_bytes 
9223372036854771712
$ cat /sys/fs/cgroup/memory/slurm/uid_3119/job_13759394/step_batch/task_0/memory.limit_in_bytes 
9223372036854771712
$ cat /sys/fs/cgroup/memory/slurm/uid_3119/job_13759394/memory.limit_in_bytes 2147483648
$ echo $SLURM_MEM_PER_NODE
2048
```

**Comment by ronblum** [comment]
@multimeric Thanks for all the info! I'll keep this one open as a separate issue, at least for now.

**Comment by timtmok** [comment]
Slurm isn't an officially supported configuration and we may not be able to fix this in a timeline that works for you. However, we would welcome a pull request to address this.

**Comment by multimeric** [comment]
> Slurm isn't an officially supported configuration with RStudio Server and we may not be able to fix this in a timeline that works for you. However, we would welcome a pull request to address this.

To be clear, this isn't just a Slurm-specific issue. Any hierarchical cgroup, where a memory limit exists on an ancestor of the RStudio cgroup, will not work with the current implementation.

Also just noting that this issue seems to also exist with cgroups v2, according to our recent testing.
